app/models/user.rb

```diff
@@ -0,0 +1,59 @@
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :registerable, :recoverable, :rememberable, :trackable, :validatable, :omniauthable, :omniauth_providers => [:facebook, :github]
+
+  has_one :profile, dependent: :destroy
+  # has_many :profiles
+  has_many :posts, dependent: :destroy
+  has_many :comments, dependent: :destroy
+  has_many :likes, dependent: :destroy
+  has_many :sent_requests, foreign_key: :requester_id, class_name: 'Request', dependent: :destroy
+  has_many :received_requests, foreign_key: :requestee_id, class_name: 'Request', dependent: :destroy
+  has_many :requestees, through: :sent_requests, dependent: :destroy
+  has_many :requesters, through: :received_requests, dependent: :destroy
+
+  after_create :make_profile
+
+
+  def friends
+     @friends = self.requesters.where('requests.accepted = ?', 1) |
+                self.requestees.where('requests.accepted = ?', 1)
+     @friends.sort do |a, b|
+        a.created_at <=> b.created_at
+     end
```

Comment:
We need a way to do it with one nice DB query. Doing it in memory will fail sooner or later (rather sooner).
With database querying and sorting it has a chance to be super fast:)

Actually there are a few ways to deal with friendships but an approach with 2 rows per one friendships seems to beat all others. See https://stackoverflow.com/questions/4219979/sql-best-practice-for-a-friendship-table and  this one https://explainextended.com/2009/03/07/selecting-friends/ (the latter is must-read 😍).

In short: we can create to rows for each friendship - so lets say we have user X and Y.

row 1: requester_id: X, requestee_id Y
row 2: requestee_id: Y, requester_id: X.

In that case: first row is created when request is sent and second when request is accepted.
We will need to remember to create second row every time the request is marked as accepted.

But we will be able to define all friends relationship as:

```
has_many :friends, through: : accepted_sent_requests
has_many :accepted_sent_requests, -> { where accepted: 1 }, foreign_key: :requester_id, class_name: 'Request'
``` .

This is a solution base on you current tables.

I also found a guide which is extending this concept by adding additional table https://dankim.io/mutual-friendship-rails/. You can try it out if it looks more clear for you:)

Reply:
We are struggling with implementing the two-row solution.

In the 'must-read' article, the first query (using the two-row method) appears to be selecting all the *requested* friends of a given user, not just those who have accepted (by reciprocating the request). A list of actual friends can be gotten through a self-join select query, such as:

```sql
SELECT f1.requestee_id
FROM requests AS f1
JOIN requests AS f2
ON f1.requester_id = f2.requestee_id AND f1.requestee_id = f2.requester_id
WHERE f1.requester_id = 4
```
(compare http://sqlfiddle.com/#!9/0b0922/1 and http://sqlfiddle.com/#!15/0b092/3)

We have managed to implement this as raw SQL using our current #friends and #friend_ids methods, but are not sure how to do it in a more Rails-y way. Also, selecting the 'pending requests' is giving us trouble.

Reply:
The goal here is to make sql query as simple as possible. In exchange we need additional logic during friendship's acceptance. _must read_ article only explains performance of both approaches, it does not give you final logic of your app.

You still have _status_ columns in your table. So, if accepting friendship do sth like (pseudocode):

```
def accept
  self.accepted = 1 #(I am not sure if 1 means accepted. it could be 3 as well)
  Request.create!(requester_id: self. requestee_id, requestee_id: self.requester_id)
end

```
you will be able to use query:

```
select * from requests where requester_id = 4 and accepted = 1;
```

That's why I suggested:

```
has_many :friends, through: : accepted_sent_requests
has_many :accepted_sent_requests, -> { where accepted: 1 }, foreign_key: :requester_id, class_name: 'Request'
```

This article https://dankim.io/mutual-friendship-rails/ extends this idea. Instead of creating second row in requests table and changing status, they create create 2 rows in Friendships once request is accepted.

I am fine with both approaches, as long as you understand what is going on in the code:)